Checkout: Show introductory offer details in discount list

sirbrillig opened the pull request:
## Proposed Changes

This PR updates the sidebar discount list to include details for Introductory offers. Instead of just reading "Introductory offer", it will now read something like "Discount for first year".

Here's a Jetpack introductory offer:

Before             |  After
:-------------------------:|:-------------------------:
<img width="335" alt="Screenshot 2024-02-05 at 12 10 35 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/c2b91172-d61e-48aa-bb6f-658c3b19d350"> | <img width="332" alt="Screenshot 2024-02-05 at 12 10 56 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/977b480d-2642-429a-9777-e24211479aad">

Here's a WPCom introductory offer (note that we also merge Sale coupons, which we should have done originally in https://github.com/Automattic/wp-calypso/pull/86371):

Before             |  After
:-------------------------:|:-------------------------:
<img width="321" alt="Screenshot 2024-02-05 at 12 20 45 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/333b4be9-e452-4527-be11-cc472a203d91"> | <img width="314" alt="Screenshot 2024-02-05 at 1 44 55 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/4e023af7-7311-4ac4-bb15-b41cdfd6fc59">


> [!NOTE]
> This preserves the multi-year discount for Jetpack biennial and hiding introductory offers for Jetpack social added by https://github.com/Automattic/wp-calypso/pull/86353. They've been made modular so we could change or extend those features to other products more easily in the future (eg: showing multi-year discounts for WPCom plans), although doing so is outside the scope of this PR.

> [!NOTE]
> Jetpack introductory offers showed an asterisk next to the "Introductory offer" item and then a notice at the bottom of the list that explains the offer is valid for the first term only. I've retained this behavior but fixed it to only display if the introductory offer is indeed for the first term. There are some issues with this notice (https://github.com/Automattic/wp-calypso/issues/87194) but they are outside the scope of this PR.

Here's Jetpack with a multi-year discount:

Before             |  After
:-------------------------:|:-------------------------:
<img width="318" alt="Screenshot 2024-02-05 at 12 12 26 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/eef2176a-b9d4-486e-846b-98a31dab3d9c"> | <img width="314" alt="Screenshot 2024-02-05 at 12 12 36 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/ea9cd582-ffef-43ac-87c1-0da84599302e">

Fixes https://github.com/Automattic/wp-calypso/issues/87108

## Testing Instructions

- Add products to your cart with introductory offers (eg: most Jetpack products, or Titan mail).
- Verify that the subtotals and totals appear the same.
- Verify that the sidebar discount list shows details about each offer in a way that makes sense.
- Verify that the price in the sidebar for those offers is correct.

sirbrillig committed:
Add details to IntroductoryOfferTerms type

sirbrillig committed:
Show "Discount for first year" instead of "Introductory offer"

sirbrillig committed:
Also handle transition_after_renewal_count

sirbrillig committed:
Update translations to use existing ones and remove unlikely units

sirbrillig committed:
Show multi-year discount for all products

sirbrillig committed:
Use getMultiYearDiscountForProduct to DRY overrides

sirbrillig committed:
Further DRY multi-years in filterAndGroupCostOverridesForDisplay

sirbrillig committed:
Add framework for showing cost at end of intro offer

sirbrillig committed:
Only show multi-year discount for Jetpack biennial as before

sirbrillig committed:
Fix getMultiYearDiscountForProduct for number of years

sirbrillig committed:
Restore hiding intro offer discount for Jetpack social

sirbrillig committed:
Fix formatting and style of IntroductoryOfferFutureTotal

sirbrillig committed:
Update IntroductoryOfferFutureTotal to have correct date and price

sirbrillig committed:
Restore other cost overrides which got removed accidentally

sirbrillig committed:
Fix description of transition_after_renewal_count

sirbrillig committed:
Reuse getIntroductoryOfferIntervalDisplay for discount list

sirbrillig committed:
Show IntroductoryOfferFutureTotal for every intro offer

sirbrillig committed:
Fix key for non-grouped overrides

sirbrillig committed:
Restore "Total" text if no intro offers exist

sirbrillig committed:
Clarify math of getMultiYearDiscountForProduct

sirbrillig committed:
Show Jetpack biennial multi-year discount out of one year intro offer

sirbrillig committed:
Restore old introductory_offer_terms

sirbrillig committed:
Do not add multi-year discount to subtotal when incorrect

sirbrillig commented:
I thought of two additional things this PR should handle:

- It should merge introductory offers which have the same text, eg: "Discount for first 3 months". (As a side benefit we might also want to merge Sale coupons which have the same text, eg: "Sale: .blog Domain registration".)
- The "Due [renewal date]: [amount]" total should also include any non-introductory offer renewal costs due on that date and we may want to add an additional "Due" line for any non-introductory offers that have a renewal before (or after?) that date.

jjchrisdiehl commented:
> 'Here's two introductory offers with the same renewal date':
![image](https://github.com/Automattic/wp-calypso/assets/16580129/6bd25461-8fea-48ca-859e-b081bfa0ec22)

If these two introductory offers renew on May 1st, wouldn't the total be well over the listed $52.78? Since the discounts for these products are -$35 each, I presume total together for renewl would be north of $100 🤔 

> It should merge introductory offers which have the same text, eg: "Discount for first 3 months". (As a side benefit we might also want to merge Sale coupons which have the same text, eg: "Sale: .blog Domain registration".)

I like this idea, I was also thinking that we may want to provide some sort of symbolic link between the product and its introductory offer like so (ignore the .blog/com mismatch, its too difficult to wrangle with Calypso refresh on focus 🤣) :

<img width="414" alt="image" src="https://github.com/Automattic/wp-calypso/assets/16580129/3558c8a6-ed18-49ea-815e-c7af819d515c">

> The "Due [renewal date]: [amount]" total should also include any non-introductory offer renewal costs due on that date and we may want to add an additional "Due" line for any non-introductory offers that have a renewal before (or after?) that date.

Hmm, I think in most cases this would be fine since there would only be perhaps one or two items due, but I wonder if listing a large number of 'due items' will make the sidebar very cluttered? If anything, the ToS will already display all this (and also be very cluttered) so it may be fine just leaving those notes there?

sirbrillig committed:
Also group sale coupons and intro offers with same title

sirbrillig committed:
DRY intro offer grouping

sirbrillig committed:
Remove uniqueId as it is no longer necessary after grouping

sirbrillig committed:
Remove "Due X: Y" text for introductory offers

sirbrillig committed:
Allow showing intro offer notice for non-jetpack products

The previous hasIntroductoryDiscount was misleadingly named and located
in a file that didn't really need it. This way it is closer to where it
is used and is more clear about what it does.

Previously it only returned true for Jetpack products, but now it allows
all products with intro offers. Previously it would also show "discount
for first term" even if the intro offer applied to multiple terms, but
now it will only show if the offer is for one term.

sirbrillig committed:
Add asterisk to single term intro offers

sirbrillig commented:
> Hmm, I think in most cases this would be fine since there would only be perhaps one or two items due, but I wonder if listing a large number of 'due items' will make the sidebar very cluttered? If anything, the ToS will already display all this (and also be very cluttered) so it may be fine just leaving those notes there?

@jjchrisdiehl yeah, you're right. I've removed the whole "Due X: Y" text from this PR. We can consider improvements later.

> I like this idea, I was also thinking that we may want to provide some sort of symbolic link between the product and its introductory offer [and sale coupon]

I think this is a good idea but a little bit outside the scope of this PR. I'll make a follow-up to try it out.

sirbrillig committed:
Restrict asterisk behavior to Jetpack
